services/device-registry-file/src/main/java/org/eclipse/hono/deviceregistry/file/FileBasedRegistrationService.java

```diff
@@ -520,6 +526,102 @@ private JsonObject convertDevice(final String deviceId, final Device payload) {
         return identities.computeIfAbsent(tenantId, id -> new ConcurrentHashMap<>());
     }
 
+    /**
+     * Return all devices for a tenant.
+     *
+     * This implementation is for demo purpose, it does not support filtering and sorting.
+     * Moreover, the order of devices cannot be guaranteed by the underlying in-memory storage
+     * so the paging may be inconsistent.
+     */
+    @Override
+    public Future<OperationResult<List<DeviceWithId>>> searchDevices(final String tenantId, final Optional<Integer> limit, final Optional<Integer> offset,
+                                                                     final Optional<List<SortingOption>> sorting, final Optional<List<Filter>> filters, final Span span) {
+
+        final int actualLimit = limit.orElse(config.getDefaultPageLimit());
+        final int actualOffset = offset.orElse(0);
+
+        return Future.succeededFuture(doSearchDevices(tenantId, actualLimit, actualOffset, sorting.orElse(null), filters.orElse(null)));
+    }
+
+    private OperationResult<List<DeviceWithId>> doSearchDevices(final String tenantId, final int limit, final int offset,
+                                                                final List<SortingOption> sort, final List<Filter> filters) {
+
+        final Set allDevicesIds = getDevicesForTenant(tenantId).keySet();
+        List<DeviceWithId> devicesList = new ArrayList<>();
+
+        for (int i = offset; i < limit + offset; i++ ) {
+            final String id = (String) allDevicesIds.toArray()[i];
+            final Device device = getDevicesForTenant(tenantId).get(id).getValue();
+            final DeviceWithId withId = DeviceWithId.fromDevice(device, id);
+
+            devicesList.add(withId);
+        }
+
+        for (Filter f : filters) {
+            devicesList = filterDeviceList(devicesList, f);
+        }
+
+        for (SortingOption s : sort) {
+            devicesList = sortDeviceList(devicesList, s);
+        }
+
+        return OperationResult.ok(HttpURLConnection.HTTP_OK,
+                devicesList,
+                Optional.ofNullable(DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())),
+                Optional.empty());
+    }
+
+    private List<DeviceWithId> filterDeviceList(final List<DeviceWithId> list, final Filter filter) {
+
+        return list
+                .stream()
+                .filter(buildJsonBasedPredicate(filter))
```

Comment:
I think it would be better to combine multiple filters into a single predicate and apply it only once, rather than iterating collection multiple times. This might not be the topic for this PR, but right now we're supporting only applying all filters. By combining predicated we could support OR functionality, but this might be left for the future.

Reply:
Done :)